action_plugins/launch_app/LaunchAppAction.qml

```
@@ -0,0 +1,77 @@
+// -*- coding: utf-8; -*-
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+
+import QtQuick
+import QtQuick.Controls
+import QtQuick.Layouts
+import QtQuick.Window
+import QtQuick.Dialogs
+
+import QtQuick.Controls.Universal
+
+import Gremlin.Profile
+import Gremlin.ActionPlugins
+import "../../qml"
+
+
+Item {
+    property LaunchAppModel action
+
+    implicitHeight: _content.height
+
+    RowLayout {
+        id: _content
+
+        anchors.left: parent.left
+        anchors.right: parent.right
+
+        Label {
+            id: _label
+
+            Layout.preferredWidth: 150
+
+            text: "Application to launch"
+        }
+
+        TextField {
```

Question here is maybe if you want to allow the user to manually modify the executable or not. If not the text field can be made non-editable. If on the other hand they should edit it, maybe we can use a verification callback slot in the model with `file_exists_and_is_accessible` to ensure the final file is valid.

---

Updated TextField to trigger action updates only onEditingFinished and fixed FileDialog.onAccepted event to set action.application_path.
Also added throwing GremlinError in _set_application_path if file does not exist, but it shows in console only, not a pop-up as it would be nice.
Any suggestions?

---

I'd like to give the user ability to edit - makes it easier to copy-paste stuff, so you don't have to edit xml every time.
Let me know how you want it done and I'll modify it, if needed.

Updated TextField to trigger action updates only onEditingFinished and fixed FileDialog.onAccepted event to set action.application_path.
Also added throwing GremlinError in _set_application_path if file does not exist, but it shows in console only, not a pop-up as it would be nice.
Any suggestions?

---

That makes sense and I agree that it makes sense to let users modify by hand the content. I had hoped that `TextLField` had a good validation system, but the one it has https://doc.qt.io/qt-6/qml-qtquick-textinput.html#validator-prop is limited to a few presets that are not useful to us. So the way I'd do this is:
1. define a property like `isInputValid` in the model which checks if the file exists
2. change the outline based on that boolean, e.g. red if not valid and blue or nothing when it is valid
3. whenever the content changes the connections via signals should let QML know the value changed and thus update the UI
That way you don't have to throw an exception, which would be awkward while editing, but the user would still get a visual indication that something is wrong, since we're all trained to spot "red fields"

The whole exception and popup stuff used to work but it's misbehaving but I haven't looked into it and it may also only misbehave when run via the IDE but be ok when run as an application. Though I wouldn't inform the user about this issue with an exception as that'd get really annoying quickly. To me exceptions are really "something is brokenly wrong", like vJoy device missing etc.

---

I would not want to throw the exception either, but just some kind of pop-up or something to alert the user that something is wrong. I like red/blue outline, but I'm afraid users might not see it unless it's blinking or so, and it's so easy to "go to the next" button then realizing the old one didn't save your setting.

---

Imo, there is only so much helping one can do. If someone can't see a clearly marked error area, that's on them; I wouldn't want to subject competent users to annoying modal windows because of that. Exceptions in Gremlin are really for things that can't possibly work without being addressed, whereas here, it just won't execute a program.

Something down the road to have is a display/report thing that much more clearly shows/highlights issues with a configuration so that people have a way to see issues. Easiest instance of that is a warning sign on the inputs (again not all are visible all the time) and better is some side panel that shows error things like most IDEs do these days.